---
title: "Intro_to_MAF"
date: "2024-01-19"
---

***Deliverables***
-Upload this R Notebook to your GitHub and submit the link to your Repo on BB.
-Include any graphs or figures created in this assignment in the folder with your R notebook with descriptive file names.

Since this is an optional partner activity, it is okay if your answers are the same as your partner’s as long as everyone understands it and could explain it in their own words if asked. Each person must individually push their code to Github. *At the top of your R Notebook, write the name of you and your partner(s) as a comment.*

***Complete the following coding activity and answer any following questions as comments in your R Notebook***

In MAF_Tutorial.Rmd, you looked at how to analyze mutation data in terms of clinical variables as well as mutation status. In this assignment, you will need to combine the two skills to demonstrate your understanding of categorical variables and R data structures.

*Pre-Assignment*
Load all necessary packages, read in the clinical data .csv file you have in your analysis_data folder, and instantiate the MAF_object.
```{r}
knitr::opts_knit$set(root.dir = normalizePath("/Users/meganrajan/Documents/qbio490/qbio_490_meganrajan/analysis_data"))

library(BiocManager)
library(TCGAbiolinks)
library(maftools)

clinical <- read.csv("/Users/meganrajan/Documents/qbio490/qbio_490_meganrajan/analysis_data/skcm_clinical_data.csv")
colnames(clinical)[ colnames(clinical) == "bcr_patient_barcode" ] <- "Tumor_Sample_Barcode"
write.csv(clinical, "/Users/meganrajan/Documents/qbio490/qbio_490_meganrajan/analysis_data/skcm_clinical_data.csv", row.names = FALSE)

```

```{r}
maf_query <- GDCquery(
    project = "TCGA-SKCM", 
    data.category = "Simple Nucleotide Variation", 
    access = "open", # we only have access to somatic mutations which are open access
    data.type = "Masked Somatic Mutation", 
    workflow.type = "Aliquot Ensemble Somatic Variant Merging and Masking"
)

# GDCdownload(maf_query) # you shouldn't have to redownload any data!

maf <- GDCprepare(maf_query) # as long as it runs, ignore any errors

maf_object <- read.maf(maf = maf, 
                       clinicalData = clinical,
                       isTCGA = TRUE)
# DONE RUNNING -- NO ERRORS
```


*1*
Choose a clinical variable to separate your populations into two different groups and rewrite the column or create a new column with that variable as a factor. ***Do not use age or vital_status as your clinical variable.*** Hint: if your variable is continuous, you will need to determine your own cutoffs for the different levels of the factor. If your variable is categorical and has more than two possible values, choose the two that are the most common.
```{r}

clinical$tumor_status = as.factor(clinical$tumor_status)

write.csv(clinical, "/Users/meganrajan/Documents/qbio490/qbio_490_meganrajan/analysis_data/skcm_clinical_data_modified.csv", row.names = FALSE)

# DONE RUNNING - NO ERRORS
```


*2*
Create a co-oncoplot with the top 10-20 (you choose) most mutated genes for the two groups. Pick one that has a large discrepancy in % mutated or type of mutations between the groups and research it. 
Research it. What is the gene used for? Can you think of any reason for the discrepancy?

```{r}

maf_clinical = maf_object@clinical.data

tsb_tumor_free <- maf_clinical[maf_clinical$tumor_status == "TUMOR FREE", ]
tsb_tumor_free <- tsb_tumor_free$Tumor_Sample_Barcode

tumor_free_maf <- subsetMaf(maf = maf_object, tsb = tsb_tumor_free)

tsb_with_tumor <- maf_clinical[maf_clinical$tumor_status == "WITH TUMOR", ]
tsb_with_tumor <- tsb_with_tumor$Tumor_Sample_Barcode

with_tumor_maf <- subsetMaf(maf = maf_object, tsb = tsb_with_tumor)

coOncoplot(m1 = tumor_free_maf,
           m2 = with_tumor_maf,
           m1Name = "TUMOR FREE",
           m2Name = "WITH TUMOR",
           borderCol = NA)
```
All 5 of the genes have similar %s in the tumor free vs. with tumor groups. However, the PCLO gene has the greatest discrepancy, iwth 48% mutations in tumor free, and 41% mutations in with tumor group. After research, the PCLO gene stands fo piccolo presynaptic cytomatrix, and functions to establish active synaptic zones and in synaptic vesicle trafficking.

*3*
Create a contingency table with your variable and chosen gene. Run a Fisher’s Exact Test between presence of mutations for that gene and your clinical variable. Create and save a mosaic plot. 
Interpret the output of the Fisher’s Exact Test in terms of the odds ratio and p-value.
Rewrite the vital_status and gender columns of your clinical df as factors.

```{r Exercise 4.1}

# Extract mutation data for PCLO gene
pclo_mutations <- maf_object@data[maf_object@data$Hugo_Symbol == "PCLO",]

clinical$PCLO_mutation <- ifelse(clinical$Tumor_Sample_Barcode %in% pclo_mutations$Tumor_Sample_Barcode, "Mutated", "Wild Type")
clinical$PCLO_mutation <- as.factor(clinical$PCLO_mutation)

# contingency table for tumor_status and PCLO_mutation
contingency_table <- table(clinical$tumor_status, clinical$PCLO_mutation)

fishers_test_result <- fisher.test(contingency_table)

clinical$vital_status <- as.factor(clinical$vital_status)
clinical$gender <- as.factor(clinical$gender)

write.csv(clinical, "/Users/meganrajan/Documents/qbio490/qbio_490_meganrajan/analysis_data/skcm_clinical_data_modified_factors.csv", row.names = FALSE)

mosaicplot(contingency_table, main="Mosaic Plot of Tumor Status vs. PCLO Mutation", color=TRUE)

```

*4*
Subset your maf_object based on your chosen clinical variable and create a co-lollipop plot of your chosen gene divided between the two different clinical variable possibilities. Include descriptive names on your plot.
Do you notice any difference in terms of mutations (e.g. sites, types, number) between the two populations?
```{r}

lollipopPlot(maf_object, gene = "PCLO")

lollipopPlot2(m1 = with_tumor_maf,
           m2 = tumor_free_maf,
           m1_name = "WITH TUMOR",
           m2_name = "TUMOR FREE",
           gene = "PCLO")

```


*5*
Create your Overall_Survival_Status column and create a mafSurvival KM plot based on mutations in your chosen gene.
Does there seem to be a difference? Hypothesize why or not based on the other analysis you did with the gene above.
```{r}
library(survminer)
library(survival)

clinical$death_days_to <- ifelse(!is.na(clinical$death_days_to), clinical$death_days_to, clinical$last_contact_days_to)
clinical$event <- ifelse(clinical$vital_status == "Dead", 1, 0)  # 

clinical$Overall_Survival_Status <- with(clinical, paste(event, birth_days_to, sep="_"))

maf_object <- read.maf(maf = maf, clinicalData = clinical, isTCGA = TRUE)

mafSurvival(maf = maf_object, genes = 'PCLO', time = 'birth_days_to', Status = 'event', isTCGA = TRUE)

```

The KM doesn't show a significant difference in survival between the mutant and wild-type groups. I am hypothesizing this due to the overlapping survival curves and the high p-value of 0.32. This suggests that mutations in the gene analyzed do not substantially impact overall survival. This makes sense with the previous analysis where the mutation rate discrepancy in the PCLO gene did not correlate with tumor status, indicating that PCLO mutations may not be a critical factor in survival outcomes.




